ibis/common/scope.py

```diff
@@ -0,0 +1,140 @@
+""" Module for scope
+    The motivation of Scope is to cache data for calculated ops.
+
+    `scope` in Scope class is the main cache. It is a dictionary mapping
+    ibis node instances to concrete data, and the time context associate
+    with it(if any).
+
+    When there are no time contexts associate with the cached result, getting
+    and setting values in Scope would be as simple as get and set in a normal
+    dictonary. With time contexts, we need the following logic for getting
+    and setting items in scope:
+
+    Set scope kv pair: before setting the value op in scope we need to perform
+    the following check first:
+
+    Test if op is in scope yet
+    - No, then put op in scope, set timecontext to be the current timecontext
+    (None if timecontext is not present), set value to be the DataFrame or
+    Series of the actual data.
+    - Yes, then get the timecontext stored in scope for op as old_timecontext,
+    and compare it with current timecontext:
+    If current timecontext is a subset of old timecontext, that means we
+    already cached a larger range of data. Do nothing and we will trim data in
+    later execution process.
+    If current timecontext is a superset of old timecontext, that means we
+    need to update cache. Set value to be the current data and set timecontext
+    to be the current timecontext for op.
+    If current timecontext is neither a subset nor a superset of old
+    timcontext, but they overlap, or not overlap at all. For example this will
+    happen when there is a window that looks forward, over a window that looks
+    back. So in this case, we should not trust the data stored either, and go
+    on to execute this node. For simplicity, we update cache in this case as
+    well.
+
+"""
+from typing import Optional
+
+from ibis.expr.timecontext import TimeContextRelation, compare_timecontext
+from ibis.expr.typing import TimeContext
+
+
+class Scope:
```

Comment:
Sorry didn't change this earlier. I think it might make more sense to define this under `ibis.expr.scope` instead of `ibis.common.scope`. Because this is the expression result caching mechanism.

Reply:
+1